File: frontend/src/types/story.ts
// Story documents (product reshape §2 / P1.1). A story is a DERIVED view over a
// conversation's canon turns — content is never stored on the story row.

export interface StoryBeat {
  turn_id: string;   // the beat's message_id (shared identity with the chat editor)
  text: string;
}

export interface StorySummary {
  id: string;
  source_conversation_id: string;
  title: string | null;
  pinned: boolean;
  cover_state: Record<string, unknown>;
  color_map: StoryColorMap;
  created_at: string | null;
  updated_at: string | null;
}

export interface StoryDetail extends StorySummary {
  beats: StoryBeat[];
}

// stories.color_map (piece tints + speaker colors). Optional/forward-compatible:
// beat_tints maps a beat's turn_id → a CSS color for its piece tint. Empty until the
// (deferred) tint writer populates it.
export interface StoryColorMap {
  beat_tints?: Record<string, string>;
  [k: string]: unknown;
}

File: backend/src/schemas/story.py
"""Story document schemas (product reshape §2 / P1.1). The story row is metadata only;
its content is DERIVED from the source conversation's canon turns at read time."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class StoryCreate(BaseModel):
    source_conversation_id: UUID
    title: str | None = None


class StoryMetaPatch(BaseModel):
    title: str | None = None
    pinned: bool | None = None
    cover_state: dict | None = None


class StoryBeat(BaseModel):
    """One canon beat, back-referencing its turn (for edit / reopen-in-chat)."""
    turn_id: UUID
    text: str


class StorySummary(BaseModel):
    id: UUID
    source_conversation_id: UUID
    title: str | None = None
    pinned: bool = False
    cover_state: dict = Field(default_factory=dict)
    color_map: dict = Field(default_factory=dict)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class StoryDetail(StorySummary):
    beats: list[StoryBeat] = Field(default_factory=list)
